### DISCUSSION TITLE
How to suppress [ccls 32] unused function 'function name' [E]?

### DISCUSSION THREAD

**Discussion by harvey-wei** [neoclide/coc.nvim#3648] [Q&A]
On ubuntu 16.04
NVIM v0.6.1
Build type: Debug
LuaJIT 2.1.0-beta3

ccls version 0.20210330-0-g0ada56ef
clang version 7.0.0-3~ubuntu0.18.04.1 (tags/RELEASE_700/final)

**Comment by chemzqm** [comment]
Can't understand

**Reply by harvey-wei** [reply]
> Can't understand

![image](https://user-images.githubusercontent.com/51208983/154406734-77b38d41-77a9-4783-a02b-98f62e3a4434.png)

If I type a new function in a .cc function, then I get the above error message. On the other hand, when I define a new function in header file, no error reporting occurs. 
After viewing the source code of ccls, I the following configuration to suppress unused-function mesage:

![image](https://user-images.githubusercontent.com/51208983/154407323-6f4f1592-fca3-42ed-8de9-c51d1aeef331.png)
https://github.com/MaskRay/ccls/blob/5a48e6c4197ee3c697d42d2e5c1d8fbab487aa36/src/sema_manager.cc

**Reply by chemzqm** [reply]
"diagnostic.showUnused" configuration might help

**Reply by harvey-wei** [reply]
> "diagnostic.showUnused" configuration might help

Nice! It works! But how do you know such configurations?

**Reply by harvey-wei** [reply]
> "diagnostic.showUnused" configuration might help

![image](https://user-images.githubusercontent.com/51208983/154415266-5ae90004-bc12-48a3-8788-f05563ec8220.png)
But it still fails for static inline function.

**Reply by chemzqm** [reply]
Try completion by coc-json or read the help

**Reply by harvey-wei** [reply]
I can not find the "diagnostic.showUnused" in coc-nvim.txt or in ccls help documentation.

> Try completion by coc-json or read the help

**Reply by chemzqm** [reply]
That was a mistake.

**Reply by harvey-wei** [reply]
> That was a mistake.

Hi! Where should I find the related configs? And should put diagnostics.showUnused in clang part of coc-settings.json like:
{
  "languageserver": {
    "ccls": {
      "command": "ccls",
      "filetypes": ["c", "cpp", "cuda", "objc", "objcpp"],
      "rootPatterns": [".ccls-root", "build/compile_commands.json"],
      "initializationOptions": {
        "cache": {
          "directory": ".ccls-cache"
        },
        "client": {
          "snippetSupport": true
        },
        "compilationDatabaseDirectory": "build",
        "clang": { 
          "extraArgs": [ "-std=c++20" ] ,
        "diagnostic":{
         "showUnused": false
         }
        }
      }
    }
  }
}

**Reply by chemzqm** [reply]
https://github.com/neoclide/coc.nvim/wiki/Using-the-configuration-file

**Reply by harvey-wei** [reply]
> https://github.com/neoclide/coc.nvim/wiki/Using-the-configuration-file

![image](https://user-images.githubusercontent.com/51208983/154629821-25c52ce8-c239-42ae-9277-b181cb3b8fbf.png)
It still can not suppress the _unused function message_  for static inline function!

**Reply by chemzqm** [reply]
Send your feature to ccls for report the diagnostic with unused tag.

**Reply by harvey-wei** [reply]
> Send your feature to ccls for report the diagnostic with unused tag.

Hi! I switch to coc-clangd. But "diagnostic.showUnused" still did not work. I guess the cause might be the -Wunused in compile_comamnds.json. Is there any way to override the clang flags?
